doc/api/esm.md

```diff
@@ -313,50 +277,193 @@ If a package has no exports, setting `"exports": false` can be used instead of
+sources in their package, with `package.json` `"main"` specifying the CommonJS
+entry point and `package.json` `"module"` specifying the ES module entry point.
+This enabled Node.js to run the CommonJS entry point while build tools such as
+bundlers used the ES module entry point, since Node.js ignored (and still
+ignores) `"module"`.
+
+In addition, there is a hazard in doing this in that the `pkg` variable created
+from `import pkg from 'pkg'` is not the same singleton as the `pkg` variable
+created from `const pkg = require('pkg')`, so if both are referenced within the
+same app (including dependencies), unexpected behavior might occur.
+
+There are three general approaches to addressing this limitation while still
+publishing a package that caters to both CommonJS and ES module consumers:
+
+1. Write the package in CommonJS or transpile the package into CommonJS, and
+  create an ES module wrapper file to provide support for named exports (for
+  example `import { name } from 'pkg'` instead of
+  `import pkg from 'pkg'; pkg.name`). Using conditional exports, the ES module
+  wrapper is used for `import` and the CommonJS entry point for `require`,
+  without introducing the hazard described above.
+
+    _pkg/package.json_
+    ```json
+    {
+      "exports": {
+        ".": {
+          "require": "./index.cjs",
+          "default": "./wrapper.mjs"
+        }
+      }
+    }
+    ```
+
+    _pkg/index.cjs_
+    ```js
+    exports.name = 'value';
+    ```
+
+    _pkg/wrapper.mjs_
+    ```js
+    import cjsModule from './index.cjs';
+    export const name = cjsModule.name;
+    ```
+
+1. Document a new ES module entry point that’s not the package `"main"`, e.g.
+  `import pkg from 'pkg/module'`. The package `"main"` would still point to
+  a CommonJS file, and thus the package would remain compatible with older
+  versions of Node.js that lack support for ES modules.
+
+    _pkg/package.json_
+    ```json
+    {
+      "type": "module",
+      "main": "./index.cjs",
+      "exports": {
+        "./module": "./index.js"
+      }
+    }
+    ```
+
+1. Set the package `"main"` entry point to an ES module file. This version and
+  above would only be usable in ES module-supporting versions of Node.js. To
+  provide access for CommonJS consumers, document a separate
+  `require('pkg/commonjs')` entry point.
+
+    _pkg/package.json_
+    ```json
+    {
+      "type": "module",
+      "main": "./index.js"
+    }
+    ```
+
+Of course, a package could also include only CommonJS or only ES module sources.
+An existing package could make a breaking change version bump to an ES
+module-only version, that would only be supported in ES module-supporting
+versions of Node.js (and other runtimes). New packages could be published
+containing only ES module sources, and would be compatible only with ES
+module-supporting runtimes.
+
+It is strongly discouraged to publish packages that use conditional exports
```

Comment:
I don't agree with "strongly discouraged" - need we editorialize here? I, for example, strongly discourage ever publishing a package that is not backwards-compatible with CJS and with pre-ESM node, but I don't think that opinion belongs in the docs either.

Reply:
I agree that in general the docs should be neutral, but this is a bit of a special case. Clearly there are folks like @MylesBorins who are concerned by this hazard, to the point where we’re considering preventing it entirely. If we’re going to let users potentially hang themselves with it, I think it’s justified to both give them guidance on what _to_ do (the first approach in this section) and what _not_ to do (this paragraph), so that they can avoid the bad pattern. Yes it’s an opinion, but it’s important for the ecosystem that this hazard doesn’t become a common pattern.

Reply:
I would argue that it's even more important that non-back-compat packages become a common pattern.

Reply:
Should we soften this to "people should be careful when doing this"? E.g. outline some general rules/pitfalls like:

* Don't use symbols as markers or make sure the symbols are shared.
* Don't use `instanceof` checks or stateful variables or make sure class declarations / stateful variables are shared.
* Carefully document the differences in behavior between the value returned by `require` and the default export of your ESM variant.

Reply:
I mean, we can express both opinions if we want 😄 They don’t need to compete against each other.

I think it’s important to express an opinion on this one because almost no users will have any idea what this hazard is and why it’s important to avoid it (or at least, to understand it and know that it’s irrelevant to your particular package).

Reply:
I’m happy to see the hazards addressed (including the minority cases when they are hazards at all) - i don’t want the hazard overblown, as i think it currently is.

Reply:
My complaint about 'strongly discouraged' is that this is a blanket statement which very often does not apply.  I agree with @jkrems suggestion that specific pitfalls should be documented instead.  These pitfalls also apply to pure CJS and pure ESM modules where multiple semver-major releases of the same module could be installed.

Partially related options 2 and 3 above (`import from 'pkg/module'` vs `require('pkg')` or the reverse) do not protect against the singleton hazard.  If one dependency uses `import from 'pkg/module'` and another dependency uses `require('pkg')` you will have split singletons (again if the hazard even applies to `pkg`).

Reply:
I have reworded this to the following:

> It is strongly encouraged to only publish CommonJS or use a wrapper approach
as described in (1) above instead of publishing packages that use conditional
exports to provide both transpiled and untranspiled versions of the same
package, e.g. with: ...

So we are only advising recommended practices. I think it is important that only package authors who understand the problem should choose to ignore this advice.

Reply:
It's definitely the same overall problem in having twice as much code in play for bundlers / two instances in the running app for singleton bugs. But in this case the consumer can at least be aware of it in debugging, instead of it being implicit and hidden behind the same specifier name.

Reply:
Marked as resolved as I've removed this docs section entirely, to just focus on the API docs changes.